### PR TITLE
allow overriding log level for retryable exception logging

### PR DESCRIPTION
**Pull request by efonsell**

### CODE REVIEW

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java [hunk 1]
```diff
@@ -195,6 +197,21 @@ private void runImpl() {
     logger.debug("Finished.");
   }
 
+  private void logRetryableException(WorkflowState state, Throwable t) {
+    switch (state.getExceptionSeverity(t)) {
+    case INFO:
+      logger.info("Handler threw a retryable exception, trying again later. Message: ", t.getMessage());
```
[the remark this answers is not in the document]
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java]: state.name() maybe? Yeah, sounds like a good addition.

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java [hunk 2]
```diff
@@ -195,6 +200,32 @@ private void runImpl() {
...
+    BiConsumer<String, Object> logMethod = getLogMethod(exceptionSeverity.logLevel);
+    if (exceptionSeverity.logStackTrace) {
+      logMethod.accept("Handler threw a retryable exception, trying again later.", t);
+    } else {
+      logMethod.accept("Handler threw a retryable exception, trying again later. Message: {}", t.getMessage());
+    }
+  }
+
+  private BiConsumer<String, Object> getLogMethod(Level logLevel) {
+    switch (logLevel) {
```
Comment (gmokki) [inline nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java:214]: Does the logger not have a generic logger method that takes the level as first argument?

**Review by gmokki — APPROVED** [review]

**Comment by gmokki** [comment]
Maybe release notes + test too...

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java (hunk 2 above)
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java:214]: According to this, no: https://stackoverflow.com/questions/2621701/setting-log-level-of-message-at-runtime-in-slf4j

File: nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java]: Added

File: nflow-engine/src/main/java/io/nflow/engine/workflow/definition/WorkflowState.java [hunk 3]
```diff
@@ -39,4 +39,16 @@ default String getDescription() {
...
+  /**
+   * Return the severity of the exception thrown by the state execution. Default is ERROR. Stack trace will not be logged for
+   * INFO.
+   *
+   * @param thrown
+   *          The thrown exception.
+   * @return Exception severity.
+   */
+  default ExceptionSeverity getExceptionSeverity(Throwable thrown) {
+    return ExceptionSeverity.ERROR;
```
[the remark this answers is not in the document]
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/workflow/definition/WorkflowState.java]: Moved to settings in v2

**Pull request closed without merging**